File: 1058-lexicographically-smallest-equivalent-string/1058-lexicographically-smallest-equivalent-string.cpp
class Solution {
public:
    vector<int> headchar;
    int find(int x){
        if(headchar[x] == -1){
            return x;
        }
        return headchar[x] = find(headchar[x]);
    }
    void Union(int x, int y){
        int parentx = find(x), parenty = find(y);
        if(parentx == parenty)return;
        headchar[max(parentx, parenty)] = min(parentx, parenty);
    }
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        headchar.resize(26,-1);
        for(int i=0; i < s1.size(); i++){
            Union(s1[i] - 'a', s2[i] - 'a');
        } 
        for(int i=0; i < baseStr.size(); i++){
            baseStr[i] = find(baseStr[i] - 'a') + 'a';
        }  
        return baseStr;
    }
};

File: 1058-lexicographically-smallest-equivalent-string/1058-lexicographically-smallest-equivalent-string.py
class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        parent = [-1] * 26

        def find(x):
            if parent[x] == -1:
                return x
            parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            px, py = find(x), find(y)
            if px == py:
                return
            parent[max(px, py)] = min(px, py)

        for a, b in zip(s1, s2):
            union(ord(a) - ord('a'), ord(b) - ord('a'))

        result = []
        for ch in baseStr:
            result.append(chr(find(ord(ch) - ord('a')) + ord('a')))

        return ''.join(result)
